cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -261,58 +315,147 @@ private void restartIfNecessary(int podId, RestartContext restartContext)
             throw new UnforceableProblem("Error getting pod " + podName(podId), e);
         }
 
-        String reasonToRestartPod = podNeedsRestart.apply(pod);
-        if (reasonToRestartPod != null && !reasonToRestartPod.isEmpty()) {
-            log.info("Pod {} needs to be restarted. Reason: {}", podId, reasonToRestartPod);
-            Admin adminClient = null;
-            try {
-                try {
-                    adminClient = adminClient(podId);
-                    Integer controller = controller(podId, adminClient, operationTimeoutMs, TimeUnit.MILLISECONDS, restartContext);
-                    int stillRunning = podToContext.reduceValuesToInt(100, v -> v.promise.future().isComplete() ? 0 : 1,
-                            0, Integer::sum);
-                    if (controller == podId && stillRunning > 1) {
-                        log.debug("Pod {} is controller and there are other pods to roll", podId);
-                        throw new ForceableProblem("Pod " + podName(podId) + " is currently the controller and there are other pods still to roll");
-                    } else {
-                        if (canRoll(adminClient, podId, 60_000, TimeUnit.MILLISECONDS)) {
-                            log.debug("Pod {} can be rolled now", podId);
+        try (RestartPlan restartPlan = restartPlan(podId, pod)) {
+            if (restartPlan.needsRestart || restartPlan.needsReconfig) {
+                if (deferController(podId, restartContext, restartPlan)) {
+                    log.debug("{}: Pod {} is controller and there are other pods to roll", reconciliation, podId);
+                    throw new ForceableProblem("Pod " + podName(podId) + " is currently the controller and there are other pods still to roll");
+                } else {
+                    if (canRoll(restartPlan.adminClient(), podId, 60_000, TimeUnit.MILLISECONDS)) {
+                        // Check for rollability before trying a dynamic update so that if the dynamic update fails we can go to a full restart
+                        if (!maybeDynamicUpdateBrokerConfig(podId, restartPlan)) {
+                            log.debug("{}: Pod {} can be rolled now", reconciliation, podId);
                             restartAndAwaitReadiness(pod, operationTimeoutMs, TimeUnit.MILLISECONDS);
                         } else {
-                            log.debug("Pod {} cannot be rolled right now", podId);
-                            throw new UnforceableProblem("Pod " + podName(podId) + " is currently not rollable");
+                            // TODO do we need some check here that the broker is still OK?
```

Comment:
What about this TODO? Should it be implemented? Or removed?

Reply:
@tombentley ^

Reply:
* I do think that we might want to have some deeper insight into whether the dynamic update has borked the broker. But it's difficult to really say for sure because I don't know what configs that are user-owned are dynamically updateable, and thus know which might come with the possibility of borkage. 
* In the longer term it's also possible we use this mechanism for certain broker-owned configs, where there's significantly higher chance of borkage.

But your question is really about the `// TODO`, not questioning what the `// TODO` is about. @stanlyDoge can you open an issue for this?

Reply:
@tombentley Concerning the TODO for this, the brokerConfig is checked during the restartPlan creation a bit before this if/else statement within the same method. Is that not sufficient for what's being asked? Are you looking for that code to be reused again in that else statement? If not, do you have an example you could point me too for some clarification on this issue?